File: Java/src/domwalker/DomFactory.java
package domwalker;

import java.util.List;

import htmlwalker.ITag;
import htmlwalker.platform.*;

public abstract class DomFactory<A extends DomApi<D>, D> extends WalkerFactory
{	
	private final WalkerPlatform platform;
	
	protected final A dapi;
	
	protected D dom;
	
	protected DomFactory(A dapi, D dom, WalkerPlatform platform)
	{
		this.dapi = dapi;
		this.dom = dom;
		this.platform = platform;
	}
	
	public abstract void close();
	
	@Override
	public WalkerPlatform platform() { return this.platform; }
	
	@Override
	public Object createTextNode(String text) { return dapi.createTextNode(dom, text); }
	
	@Override
	public Object createEntityNode(String entity) { return dapi.createTextNode(dom, entity); }

	@Override
	public List<Object> createContent(ITag adapter) { return dapi.getContent(dom, adapter); }
	
    @Override
    public Object createA() { return dapi.createElement(dom, "a"); }

    @Override
    public Object createAbbr() { return dapi.createElement(dom, "abbr"); }

	@Override
	public Object createAcronym() { return dapi.createElement(dom, "acronym"); }

    @Override
    public Object createAddress() { return dapi.createElement(dom, "address"); }

    @Override
    public Object createArea() { return dapi.createElement(dom, "area"); }

    @Override
    public Object createArticle() { return dapi.createElement(dom, "article"); }

    @Override
    public Object createAside() { return dapi.createElement(dom, "aside"); }

    @Override
    public Object createAudio() { return dapi.createElement(dom, "audio"); }

    @Override
    public Object createB() { return dapi.createElement(dom, "b"); }

    @Override
    public Object createBase() { return dapi.createElement(dom, "base"); }

    @Override
    public Object createBdi() { return dapi.createElement(dom, "bdi"); }

    @Override
    public Object createBig() { return dapi.createElement(dom, "big"); }

    @Override
    public Object createBdo() { return dapi.createElement(dom, "bdo"); }

    @Override
    public Object createBlockquote() { return dapi.createElement(dom, "blockquote"); }

    @Override
    public Object createBody() { return dapi.createElement(dom, "body"); }

    @Override
    public Object createBr() { return dapi.createElement(dom, "br"); }

    @Override
    public Object createButton() { return dapi.createElement(dom, "button"); }

    @Override
    public Object createCanvas() { return dapi.createElement(dom, "canvas"); }

    @Override
    public Object createCaption() { return dapi.createElement(dom, "caption"); }

    @Override
    public Object createCenter() { return dapi.createElement(dom, "center"); }

    @Override
    public Object createCite() { return dapi.createElement(dom, "cite"); }

    @Override
    public Object createCode() { return dapi.createElement(dom, "code"); }

    @Override
    public Object createCol() { return dapi.createElement(dom, "col"); }

    @Override
    public Object createColgroup() { return dapi.createElement(dom, "colgroup"); }

    @Override
    public Object createCommand() { return dapi.createElement(dom, "command"); }

    @Override
    public Object createDatalist() { return dapi.createElement(dom, "datalist"); }

    @Override
    public Object createDd() { return dapi.createElement(dom, "dd"); }

    @Override
    public Object createDel() { return dapi.createElement(dom, "del"); }

    @Override
    public Object createDetails() { return dapi.createElement(dom, "details"); }

    @Override
    public Object createDialog() { return dapi.createElement(dom, "dialog"); }

    @Override
    public Object createDiv() { return dapi.createElement(dom, "div"); }

    @Override
    public Object createDfn() { return dapi.createElement(dom, "dfn"); }

    @Override
    public Object createDl() { return dapi.createElement(dom, "dl"); }

    @Override
    public Object createDt() { return dapi.createElement(dom, "dt"); }

    @Override
    public Object createEm() { return dapi.createElement(dom, "em"); }

    @Override
    public Object createEmbed() { return dapi.createElement(dom, "embed"); }

    @Override
    public Object createFigcaption() { return dapi.createElement(dom, "figcaption"); }

    @Override
    public Object createFigure() { return dapi.createElement(dom, "figure"); }

    @Override
    public Object createFieldset() { return dapi.createElement(dom, "fieldset"); }

    @Override
    public Object createFooter() { return dapi.createElement(dom, "footer"); }

    @Override
    public Object createFont() { return dapi.createElement(dom, "font"); }

    @Override
    public Object createForm() { return dapi.createElement(dom, "form"); }

    @Override
    public Object createH1() { return dapi.createElement(dom, "h1"); }

    @Override
    public Object createH2() { return dapi.createElement(dom, "h2"); }

    @Override
    public Object createH3() { return dapi.createElement(dom, "h3"); }

    @Override
    public Object createH4() { return dapi.createElement(dom, "h4"); }

    @Override
    public Object createH5() { return dapi.createElement(dom, "h5"); }

    @Override
    public Object createH6() { return dapi.createElement(dom, "h6"); }

    @Override
    public Object createHead() { return dapi.createElement(dom, "head"); }

    @Override
    public Object createHeader() { return dapi.createElement(dom, "header"); }

    @Override
    public Object createHgroup() { return dapi.createElement(dom, "hgroup"); }

    @Override
    public Object createHtml() {
        Object tag = dapi.htmlElement(dom);
        if (tag == null)
            tag = dapi.createElement(dom, "html");
        return tag;
    }

    @Override
    public Object createHr() { return dapi.createElement(dom, "hr"); }

    @Override
    public Object createI() { return dapi.createElement(dom, "i"); }

    @Override
    public Object createIframe() { return dapi.createElement(dom, "iframe"); }

    @Override
    public Object createImg() { return dapi.createElement(dom, "img"); }

    @Override
    public Object createInput() { return dapi.createElement(dom, "input"); }

    @Override
    public Object createIns() { return dapi.createElement(dom, "ins"); }

    @Override
    public Object createKbd() { return dapi.createElement(dom, "kbd"); }

    @Override
    public Object createKeygen() { return dapi.createElement(dom, "keygen"); }

    @Override
    public Object createLabel() { return dapi.createElement(dom, "label"); }

    @Override
    public Object createLegend() { return dapi.createElement(dom, "legend"); }

    @Override
    public Object createLi() { return dapi.createElement(dom, "li"); }

    @Override
    public Object createLink() { return dapi.createElement(dom, "link"); }

    @Override
    public Object createMain() { return dapi.createElement(dom, "main"); }

    @Override
    public Object createMap() { return dapi.createElement(dom, "map"); }

    @Override
    public Object createMark() { return dapi.createElement(dom, "mark"); }

    @Override
    public Object createMenu() { return dapi.createElement(dom, "menu"); }

	@Override
	public Object createMenuitem() { return dapi.createElement(dom, "menuitem"); }

    @Override
    public Object createMeta() { return dapi.createElement(dom, "meta"); }

    @Override
    public Object createMeter() { return dapi.createElement(dom, "meter"); }

    @Override
    public Object createNav() { return dapi.createElement(dom, "nav"); }

    @Override
    public Object createNoscript() { return dapi.createElement(dom, "noscript"); }

    @Override
    public Object createObject() { return dapi.createElement(dom, "object"); }

    @Override
    public Object createOl() { return dapi.createElement(dom, "ol"); }

    @Override
    public Object createOptgroup() { return dapi.createElement(dom, "optgroup"); }

    @Override
    public Object createOption() { return dapi.createElement(dom, "option"); }

    @Override
    public Object createOutput() { return dapi.createElement(dom, "output"); }

    @Override
    public Object createP() { return dapi.createElement(dom, "p"); }

    @Override
    public Object createParam() { return dapi.createElement(dom, "param"); }

	@Override
	public Object createPicture() { return dapi.createElement(dom, "picture"); }

    @Override
    public Object createPre() { return dapi.createElement(dom, "pre"); }

    @Override
    public Object createProgress() { return dapi.createElement(dom, "progress"); }

    @Override
    public Object createQ() { return dapi.createElement(dom, "q"); }

    @Override
    public Object createRp() { return dapi.createElement(dom, "rp"); }

    @Override
    public Object createRt() { return dapi.createElement(dom, "rt"); }

    @Override
    public Object createRuby() { return dapi.createElement(dom, "ruby"); }

    @Override
    public Object createS() { return dapi.createElement(dom, "s"); }

    @Override
    public Object createSamp() { return dapi.createElement(dom, "samp"); }

    @Override
    public Object createScript() { return dapi.createElement(dom, "script"); }

    @Override
    public Object createSection() { return dapi.createElement(dom, "section"); }

    @Override
    public Object createSelect() { return dapi.createElement(dom, "select"); }

    @Override
    public Object createSmall() { return dapi.createElement(dom, "small"); }

    @Override
    public Object createSource() { return dapi.createElement(dom, "source"); }

    @Override
    public Object createSpan() { return dapi.createElement(dom, "span"); }

    @Override
    public Object createStrike() { return dapi.createElement(dom, "strike"); }

    @Override
    public Object createStrong() { return dapi.createElement(dom, "strong"); }

    @Override
    public Object createStyle() { return dapi.createElement(dom, "style"); }

    @Override
    public Object createSub() { return dapi.createElement(dom, "sub"); }

    @Override
    public Object createSummary() { return dapi.createElement(dom, "summary"); }

    @Override
    public Object createSup() { return dapi.createElement(dom, "sup"); }

    @Override
    public Object createTable() { return dapi.createElement(dom, "table"); }

    @Override
    public Object createTbody() { return dapi.createElement(dom, "tbody"); }

    @Override
    public Object createTd() { return dapi.createElement(dom, "td"); }

    @Override
    public Object createTextarea() { return dapi.createElement(dom, "textarea"); }

    @Override
    public Object createTfoot() { return dapi.createElement(dom, "tfoot"); }

    @Override
    public Object createTh() { return dapi.createElement(dom, "th"); }

    @Override
    public Object createThead() { return dapi.createElement(dom, "thead"); }

    @Override
    public Object createTime() { return dapi.createElement(dom, "time"); }

    @Override
    public Object createTitle() { return dapi.createElement(dom, "title"); }

    @Override
    public Object createTr() { return dapi.createElement(dom, "tr"); }

    @Override
    public Object createTrack() { return dapi.createElement(dom, "track"); }

	@Override
	public Object createTt() { return dapi.createElement(dom, "tt"); }

    @Override
    public Object createU() { return dapi.createElement(dom, "u"); }

    @Override
    public Object createUl() { return dapi.createElement(dom, "ul"); }

    @Override
    public Object createVar() { return dapi.createElement(dom, "var"); }

    @Override
    public Object createVideo() { return dapi.createElement(dom, "video"); }

    @Override
    public Object createWbr() { return dapi.createElement(dom, "wbr"); }
	
}

File: CSharp/DomWalker/DomFactory.cs
using HtmlWalker;
using HtmlWalker.Platform;
using System.Collections.Generic;

namespace DomWalker
{
    public abstract class DomFactory<A, D> : WalkerPlatform.WalkerFactory where A : DomApi<D>
    {
        internal protected DomFactory(A api, D dom, WalkerPlatform platform)
        {
            DApi = api;
            Dom = dom;
            Platform = platform;
        }

        public abstract void Close();

        internal A DApi { get; }

        internal protected D Dom { get; set; }

        public override WalkerPlatform Platform { get; }

        protected override IList<object> CreateContent(ITag adapter)
            => DApi.GetContent(Dom, adapter);

        protected override object CreateTextNode(string text) => DApi.CreateTextNode(Dom, text);

        protected override object CreateEntityNode(string entity)
            => DApi.CreateEntityNode(Dom, entity);

        #region abstract methods to create HTML tags

        protected override object CreateA() => DApi.CreateElement(Dom, "a");
        protected override object CreateAbbr() => DApi.CreateElement(Dom, "abbr");
        protected override object CreateAcronym() => DApi.CreateElement(Dom, "acronym");
        protected override object CreateAddress() => DApi.CreateElement(Dom, "address");
        protected override object CreateArea() => DApi.CreateElement(Dom, "area");
        protected override object CreateArticle() => DApi.CreateElement(Dom, "article");
        protected override object CreateAside() => DApi.CreateElement(Dom, "aside");
        protected override object CreateAudio() => DApi.CreateElement(Dom, "audio");
        protected override object CreateB() => DApi.CreateElement(Dom, "b");
        protected override object CreateBase() => DApi.CreateElement(Dom, "base");
        protected override object CreateBdi() => DApi.CreateElement(Dom, "bdi");
        protected override object CreateBdo() => DApi.CreateElement(Dom, "bdo");
        protected override object CreateBig() => DApi.CreateElement(Dom, "big");
        protected override object CreateBlockquote() => DApi.CreateElement(Dom, "blockquote");
        protected override object CreateBody() => DApi.CreateElement(Dom, "body");
        protected override object CreateBr() => DApi.CreateElement(Dom, "br");
        protected override object CreateButton() => DApi.CreateElement(Dom, "button");
        protected override object CreateCanvas() => DApi.CreateElement(Dom, "canvas");
        protected override object CreateCaption() => DApi.CreateElement(Dom, "caption");
        protected override object CreateCenter() => DApi.CreateElement(Dom, "center");
        protected override object CreateCite() => DApi.CreateElement(Dom, "cite");
        protected override object CreateCode() => DApi.CreateElement(Dom, "code");
        protected override object CreateCol() => DApi.CreateElement(Dom, "col");
        protected override object CreateColgroup() => DApi.CreateElement(Dom, "colgroup");
        protected override object CreateCommand() => DApi.CreateElement(Dom, "command");
        protected override object CreateDatalist() => DApi.CreateElement(Dom, "datalist");
        protected override object CreateDd() => DApi.CreateElement(Dom, "dd");
        protected override object CreateDel() => DApi.CreateElement(Dom, "del");
        protected override object CreateDetails() => DApi.CreateElement(Dom, "details");
        protected override object CreateDialog() => DApi.CreateElement(Dom, "dialog");
        protected override object CreateDiv() => DApi.CreateElement(Dom, "div");
        protected override object CreateDfn() => DApi.CreateElement(Dom, "dfn");
        protected override object CreateDl() => DApi.CreateElement(Dom, "dl");
        protected override object CreateDt() => DApi.CreateElement(Dom, "dt");
        protected override object CreateEm() => DApi.CreateElement(Dom, "em");
        protected override object CreateEmbed() => DApi.CreateElement(Dom, "embed");
        protected override object CreateFigcaption() => DApi.CreateElement(Dom, "figcaption");
        protected override object CreateFigure() => DApi.CreateElement(Dom, "figure");
        protected override object CreateFieldset() => DApi.CreateElement(Dom, "fieldset");
        protected override object CreateFooter() => DApi.CreateElement(Dom, "footer");
        protected override object CreateFont() => DApi.CreateElement(Dom, "font");
        protected override object CreateForm() => DApi.CreateElement(Dom, "form");
        protected override object CreateH1() => DApi.CreateElement(Dom, "h1");
        protected override object CreateH2() => DApi.CreateElement(Dom, "h2");
        protected override object CreateH3() => DApi.CreateElement(Dom, "h3");
        protected override object CreateH4() => DApi.CreateElement(Dom, "h4");
        protected override object CreateH5() => DApi.CreateElement(Dom, "h5");
        protected override object CreateH6() => DApi.CreateElement(Dom, "h6");
        protected override object CreateHead() => DApi.CreateElement(Dom, "head");
        protected override object CreateHeader() => DApi.CreateElement(Dom, "header");
        protected override object CreateHgroup() => DApi.CreateElement(Dom, "hgroup");
        protected override object CreateHtml()
        {
            var tag = DApi.HtmlElement(Dom);
            if (tag == null)
                tag = DApi.CreateElement(Dom, "html");
            return tag;
        }
        protected override object CreateHr() => DApi.CreateElement(Dom, "hr");
        protected override object CreateI() => DApi.CreateElement(Dom, "i");
        protected override object CreateIframe() => DApi.CreateElement(Dom, "iframe");
        protected override object CreateImg() => DApi.CreateElement(Dom, "img");
        protected override object CreateInput() => DApi.CreateElement(Dom, "input");
        protected override object CreateIns() => DApi.CreateElement(Dom, "ins");
        protected override object CreateKbd() => DApi.CreateElement(Dom, "kbd");
        protected override object CreateKeygen() => DApi.CreateElement(Dom, "keygen");
        protected override object CreateLabel() => DApi.CreateElement(Dom, "label");
        protected override object CreateLegend() => DApi.CreateElement(Dom, "legend");
        protected override object CreateLi() => DApi.CreateElement(Dom, "li");
        protected override object CreateLink() => DApi.CreateElement(Dom, "link");
        protected override object CreateMain() => DApi.CreateElement(Dom, "main");
        protected override object CreateMap() => DApi.CreateElement(Dom, "map");
        protected override object CreateMark() => DApi.CreateElement(Dom, "mark");
        protected override object CreateMenu() => DApi.CreateElement(Dom, "menu");
        protected override object CreateMenuitem() => DApi.CreateElement(Dom, "menuitem");
        protected override object CreateMeta() => DApi.CreateElement(Dom, "meta");
        protected override object CreateMeter() => DApi.CreateElement(Dom, "meter");
        protected override object CreateNav() => DApi.CreateElement(Dom, "nav");
        protected override object CreateNoscript() => DApi.CreateElement(Dom, "noscript");
        protected override object CreateObject() => DApi.CreateElement(Dom, "object");
        protected override object CreateOl() => DApi.CreateElement(Dom, "ol");
        protected override object CreateOptgroup() => DApi.CreateElement(Dom, "optgroup");
        protected override object CreateOption() => DApi.CreateElement(Dom, "option");
        protected override object CreateOutput() => DApi.CreateElement(Dom, "output");
        protected override object CreateP() => DApi.CreateElement(Dom, "p");
        protected override object CreateParam() => DApi.CreateElement(Dom, "param");
        protected override object CreatePicture() => DApi.CreateElement(Dom, "picture");
        protected override object CreatePre() => DApi.CreateElement(Dom, "pre");
        protected override object CreateProgress() => DApi.CreateElement(Dom, "progress");
        protected override object CreateQ() => DApi.CreateElement(Dom, "q");
        protected override object CreateRp() => DApi.CreateElement(Dom, "rp");
        protected override object CreateRt() => DApi.CreateElement(Dom, "rt");
        protected override object CreateRuby() => DApi.CreateElement(Dom, "ruby");
        protected override object CreateS() => DApi.CreateElement(Dom, "s");
        protected override object CreateSamp() => DApi.CreateElement(Dom, "samp");
        protected override object CreateScript() => DApi.CreateElement(Dom, "script");
        protected override object CreateSection() => DApi.CreateElement(Dom, "section");
        protected override object CreateSelect() => DApi.CreateElement(Dom, "select");
        protected override object CreateSmall() => DApi.CreateElement(Dom, "small");
        protected override object CreateSource() => DApi.CreateElement(Dom, "source");
        protected override object CreateSpan() => DApi.CreateElement(Dom, "span");
        protected override object CreateStrike() => DApi.CreateElement(Dom, "strike");
        protected override object CreateStrong() => DApi.CreateElement(Dom, "strong");
        protected override object CreateStyle() => DApi.CreateElement(Dom, "style");
        protected override object CreateSub() => DApi.CreateElement(Dom, "sub");
        protected override object CreateSummary() => DApi.CreateElement(Dom, "summary");
        protected override object CreateSup() => DApi.CreateElement(Dom, "sup");
        protected override object CreateTable() => DApi.CreateElement(Dom, "table");
        protected override object CreateTbody() => DApi.CreateElement(Dom, "tbody");
        protected override object CreateTd() => DApi.CreateElement(Dom, "td");
        protected override object CreateTextarea() => DApi.CreateElement(Dom, "textarea");
        protected override object CreateTfoot() => DApi.CreateElement(Dom, "tfoot");
        protected override object CreateTh() => DApi.CreateElement(Dom, "th");
        protected override object CreateThead() => DApi.CreateElement(Dom, "thead");
        protected override object CreateTime() => DApi.CreateElement(Dom, "time");
        protected override object CreateTitle() => DApi.CreateElement(Dom, "title");
        protected override object CreateTr() => DApi.CreateElement(Dom, "tr");
        protected override object CreateTrack() => DApi.CreateElement(Dom, "track");
        protected override object CreateTt() => DApi.CreateElement(Dom, "tt");
        protected override object CreateU() => DApi.CreateElement(Dom, "u");
        protected override object CreateUl() => DApi.CreateElement(Dom, "ul");
        protected override object CreateVar() => DApi.CreateElement(Dom, "var");
        protected override object CreateVideo() => DApi.CreateElement(Dom, "video");
        protected override object CreateWbr() => DApi.CreateElement(Dom, "wbr");

        #endregion
    }
}
